### PR TITLE
Index filesystem contents by filename instead of hash.

### PR DESCRIPTION
**Pull request by pchote**
This removes the requirement for package formats to expose classic/crc hashes to the filesystem code.  These hashes were included as a workaround for the mix format not storing filenames.  We now instead require that all mix files have corresponding mix database entries (true for all default mixes as well as any custom mixes with valid local mix database.  This will break any mods that intentionally write corrupted mix databases as "copy protection").

Tested that TD/RA/TS still run, and that they can still be installed from CD (however note #10529).

### CODE REVIEW

**Comment by Mailaender** [comment]
Does this give any benefit as in performance wise?

**Comment by pchote** [comment]
In principle, yes, but only after the rest of the filesystem fixes (I have more, but they are blocked on currently open prs) are merged.

**Comment by pchote** [comment]
Rebased.

Note that this introduces an important change:  If code asks for files with the wrong case then they will miss the cache and take the slow path.  #10531 fixes the only place I could find this to be an issue.

**Comment by pchote** [comment]
I have pushed some additional cleanup commits that don't make a lot of sense to review separately.  I have tested the InstallShield and D2kSoundResources changes, but am not able to test the changes to BagFile (which is a RA2-specific format).

**Comment by pchote** [comment]
Note that this this makes InstallShieldPackage / D2kSoundResources / BagFile case sensitive.

**Comment by abcdefg30** [comment]
All mods (and RA2) still load and work. :+0.75:

File: mods/cnc/mod.yaml [hunk 1]
```diff
@@ -146,13 +146,11 @@ LoadScreen: CncLoadScreen
...
 ContentInstaller:
 	TestFiles: ^Content/cnc/conquer.mix, ^Content/cnc/desert.mix, ^Content/cnc/sounds.mix, ^Content/cnc/speech.mix, ^Content/cnc/temperat.mix, ^Content/cnc/tempicnh.mix, ^Content/cnc/winter.mix
-	FilesToCopy: CONQUER.MIX, DESERT.MIX, SCORES.MIX, SOUNDS.MIX, TEMPERAT.MIX, WINTER.MIX
-	FilesToExtract: speech.mix, tempicnh.mix, transit.mix
 	PackageMirrorList: http://www.openra.net/packages/cnc-mirrors.txt
 	DiskTestFiles: conquer.mix, desert.mix, install/setup.z
 	PackageToExtractFromCD: install/setup.z
 	ExtractFilesFromCD:
-		.: speech.mix, tempicnh.mix, transit.mix
+		.: SPEECH.MIX, TEMPICNH.MIX, TRANSIT.MIX
```
Comment (Mailaender) [inline mods/cnc/mod.yaml]: This looks like it could be dangerous for case-sensitive file systems. Don't have a CD drive to test on Linux at hand, sorry.

Comment (pchote) [inline mods/cnc/mod.yaml]: These files are inside setup.z, which lists them in upper case.  The install code .ToLowerInvariants the output filename, so there is no issue for case sensitive file systems.

File: OpenRA.Game/FileSystem/FileSystem.cs [hunk 2]
```diff
@@ -107,16 +106,9 @@ void MountInner(IReadOnlyPackage package)
 		{
 			MountedPackages.Add(package);
 
-			foreach (var hash in package.ClassicHashes())
+			foreach (var hash in package.AllFileNames())
```
Comment (penev92) [inline OpenRA.Game/FileSystem/FileSystem.cs]: Nit: This is no longer a hash.

File: OpenRA.Game/FileSystem/FileSystem.cs [hunk 3]
```diff
@@ -127,6 +119,9 @@ public bool Unmount(IReadOnlyPackage mount)
 			if (MountedPackages.Contains(mount))
 				mount.Dispose();
 
+			foreach (var f in fileIndex)
+				f.Value.RemoveAll(p => p == mount);
```
Comment (penev92) [inline OpenRA.Game/FileSystem/FileSystem.cs]: Shouldn't this go *before* the `.Dispose()` above?

**Comment by pchote** [comment]
Fixed.

**Comment by penev92** [comment]
D2k CD asset installation crashes with a duplicate key exception.
I traced it to a second "LAUNCHER.BMP" being read from `setup.z` and added to `InstallShieldPackage.index`. You will need to read the directory names inside the package as well and attach those to the file names, I guess (they are currently read and ignored).

**Comment by pchote** [comment]
Fixed D2K installation from CD – both my own issues and preexisting problems on case sensitive file systems.

**Comment by abcdefg30** [comment]
> OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromCDLogic.cs:L194: [SP2100] Code line must be no longer than 180 characters (now 181).

**Comment by pchote** [comment]
Fixed.

**Comment by pchote** [comment]
Rebased.

**Comment by Mailaender** [comment]
Dune 2000 package installation works so far except videos:

```yaml
Videos:
	Briefing: a_br01_e.vqa
```

vs.

image

**Comment by Mailaender** [comment]
RA2 also still works. :white_check_mark:

**Comment by pchote** [comment]
Fixed.

**Comment by pchote** [comment]
Rebased again.

**Comment by obrakmann** [comment]
Aside from that one comment, looks fine to me. Confirmed that current installations continue to run, and that installation from both the freeware CDs as well as downloaded content still work. :+1:

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/9ab4b27f52967b594aa9891ad176e310cff6227d)

**Pull request merged**